### PR TITLE
[avmfritz] Refactoring of discovery service and internal event handling

### PR DESCRIPTION
**Pull request by cweitkamp**
- Refactoring of discovery service and internal event handling

The main purpose of this PR was to remove a superfluous polling job for background discovery as the same request was done all the time to update the channels / properties of `ThingHandler`s.

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseBridgeHandler.java [hunk 1]
```diff
@@ -222,208 +223,46 @@ public void addTemplateList(List<TemplateModel> templateList) {
...
+     * @param deviceList list of devices
      */
-    public void addDeviceList(List<AVMFritzBaseModel> deviceList) {
-        for (Thing thing : getThing().getThings()) {
+    public void onDeviceListAdded(List<AVMFritzBaseModel> deviceList) {
+        getThing().getThings().stream().forEach(thing -> {
             AVMFritzBaseThingHandler handler = (AVMFritzBaseThingHandler) thing.getHandler();
             if (handler != null) {
                 Optional<AVMFritzBaseModel> optionalDevice = deviceList.stream()
                         .filter(it -> it.getIdentifier().equals(handler.getIdentifier())).findFirst();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseBridgeHandler.java:235]: This logic could be optimized
```suggestion
        Map<String, AVMFritzBaseModel> deviceMap = deviceList.stream().collect(Collectors.toMap(it -> it.getIdentifier(), Functions.identity()));
        getThing().getThings().forEach(thing -> {
            AVMFritzBaseThingHandler handler = (AVMFritzBaseThingHandler) thing.getHandler();
            if (handler != null) {
                Optional<AVMFritzBaseModel> optionalDevice = Optional.ofNullable(deviceMap.get(handler.getIdentifier());
```

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java [hunk 2]
```diff
@@ -78,15 +94,202 @@ public void initialize() {
...
+
+    /**
+     * Updates thing properties.
+     *
+     * @param device the {@link AVMFritzBaseModel}
+     */
+    private void updateProperties(AVMFritzBaseModel device) {
+        Map<String, String> editProperties = editProperties();
+        editProperties.put(Thing.PROPERTY_FIRMWARE_VERSION, device.getFirmwareVersion());
+        if (device instanceof GroupModel && ((GroupModel) device).getGroupinfo() != null) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java:212]: `((GroupModel) device).getGroupinfo()` should be cached to a local variable

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java [hunk 3]
```diff
@@ -78,15 +94,202 @@ public void initialize() {
...
+     * @param configId ID of the configuration to be updated.
+     * @param value Value to be set.
+     */
+    private void updateThingChannelConfiguration(String channelId, String configId, Object value) {
+        Channel channel = thing.getChannel(channelId);
+        if (channel != null) {
+            Configuration config = channel.getConfiguration();
+            Configuration editConfig = new Configuration(new HashMap<>(config.getProperties()));
+            editConfig.put(configId, value);
+            config.setProperties(editConfig.getProperties());
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java:280]: Is there a reason this isn't simplified?
```suggestion
            config.put(configId, value);
```

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java [hunk 4]
```diff
@@ -78,15 +94,202 @@ public void initialize() {
         updateStatus(ThingStatus.UNKNOWN);
     }
 
+    @Override
+    public void onDeviceAdded(AVMFritzBaseModel device) {
+        // nothing to do
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java:99]: If onDeviceRemove/onDeviceGone cause the thing to go offline, wouldn't adding the device make the thing go online?

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java [hunk 5]
```diff
@@ -74,41 +81,123 @@ public Powerline546EHandler(Bridge bridge, HttpClient httpClient,
...
+        if (callback != null) {
+            ChannelUID channelUID = new ChannelUID(thing.getUID(), channelId);
+            ChannelTypeUID channelTypeUID = CHANNEL_BATTERY.equals(channelId)
+                    ? new ChannelTypeUID("system:battery-level")
+                    : new ChannelTypeUID(BINDING_ID, channelId);
+            Channel channel = callback.createChannelBuilder(channelUID, channelTypeUID).build();
+            updateThing(editThing().withoutChannel(channelUID).withChannel(channel).build());
         }
-        super.updateThingFromDevice(thing, device);
+    }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java:191]: This logic is also used in AVMFritzBaseThingHandler so it should be refactored into a utility class.

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java [hunk 6]
```diff
@@ -74,41 +81,123 @@ public Powerline546EHandler(Bridge bridge, HttpClient httpClient,
...
+     * Creates new channels for the thing.
+     *
+     * @param channelId ID of the channel to be created.
+     */
+    private void createChannel(String channelId) {
+        ThingHandlerCallback callback = getCallback();
+        if (callback != null) {
+            ChannelUID channelUID = new ChannelUID(thing.getUID(), channelId);
+            ChannelTypeUID channelTypeUID = CHANNEL_BATTERY.equals(channelId)
+                    ? new ChannelTypeUID("system:battery-level")
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java:186]: Is there no constant in the core for this channel type?

Comment (cweitkamp) [reply bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java:186]: No, there are no constants for the channel type uids.

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java (hunk 4 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java:99]: No, `onDeviceAdded()` will only be called if no Thing is available.

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java (hunk 2 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java:249]: Yes, I am aware of all those minor improvements. To be honest, I am not very happy with the current code. My plan is to refactor the binding step by step.

**Comment by cpmeister** [comment]
So is this ready to merge or are you making more changes?

File: bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java [hunk 7]
```diff
@@ -74,41 +83,137 @@ public Powerline546EHandler(Bridge bridge, HttpClient httpClient,
...
-     * @param thing Thing to be updated.
-     * @param device Device model with new data.
+     * @param channelId ID of the channel to be created.
      */
+    private void createChannel(String channelId) {
+        ThingHandlerCallback callback = getCallback();
+        if (callback != null) {
+            ChannelUID channelUID = new ChannelUID(thing.getUID(), channelId);
+            ChannelTypeUID channelTypeUID = CHANNEL_BATTERY.equals(channelId)
+                    ? new ChannelTypeUID("system", "battery-level")
```
Comment (cpmeister) [inline bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java:205]: I found the constant for this!
`DefaultSystemChannelTypeProvider.SYSTEM_CHANNEL_BATTERY_LEVEL.getUID()`

Comment (cweitkamp) [reply bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java:205]: Nice :wink:

To make you happy i submitted https://github.com/openhab/openhab-addons/pull/7650.

**Pull request merged**